[DTensor] Fix unsupported op error

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #170889

There is already an 'unsupported op' error, but it happens late enough
during sharding prop flow that for ops that take TupleStrategy
(tensorlist) inputs, we hit a different error (missing DeviceMesh from
dtensor args) which is a confusing error to tell a user when in fact the
problem is the operator is not registered with DTensor dispatch in the
first place.

This PR attempts to fix the missing devicemesh issue so that the
pre-existing error gets hit for cat as well.

Done by claude, needs careful review!!

Claude's summary:
```
 I fixed a bug where DTensor ops with list[Tensor] arguments (like torch.cat) were raising "found no DeviceMesh from dtensor args" instead of the correct "does not
  have a sharding strategy registered" error.

  Changes made:

  1. C++ changes in torch/csrc/autograd/python_variable.cpp:
  - Modified the NON_TENSOR case in args processing (line 2245-2263) to check if a list contains DTensors and extract compute_mesh from them
  - Modified the NON_TENSOR case in kwargs processing (line 2305-2321) with the same logic

  2. Python changes in torch/distributed/tensor/_dispatch.py:
  - Modified list/tuple handling in args processing (line 550-576) to properly unwrap DTensors inside lists, extracting their specs for the schema and local tensors
  for execution
  - Modified list/tuple handling in kwargs processing (line 599-625) with the same unwrapping logic

  The fix ensures that:
  - The C++ dispatch layer finds the DeviceMesh from DTensors inside list arguments
  - The Python dispatch layer properly unwraps DTensors in lists before creating the OpSchema
  - Sharding propagation can proceed and correctly raise NotImplementedError for unsupported ops, regardless of whether they have list arguments
```

wconstab committed:
[DTensor] Fix unsupported op error

There is already an 'unsupported op' error, but it happens late enough
during sharding prop flow that for ops that take TupleStrategy
(tensorlist) inputs, we hit a different error (missing DeviceMesh from
dtensor args) which is a confusing error to tell a user when in fact the
problem is the operator is not registered with DTensor dispatch in the
first place.

This PR attempts to fix the missing devicemesh issue so that the
pre-existing error gets hit for cat as well.

Done by claude, needs careful review!!

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Fix unsupported op error"


There is already an 'unsupported op' error, but it happens late enough
during sharding prop flow that for ops that take TupleStrategy
(tensorlist) inputs, we hit a different error (missing DeviceMesh from
dtensor args) which is a confusing error to tell a user when in fact the
problem is the operator is not registered with DTensor dispatch in the
first place.

This PR attempts to fix the missing devicemesh issue so that the
pre-existing error gets hit for cat as well.

Done by claude, needs careful review!!

Claude's summary:

 I fixed a bug where DTensor ops with list[Tensor] arguments (like torch.cat) were raising "found no DeviceMesh from dtensor args" instead of the correct "does not
  have a sharding strategy registered" error.

  Changes made:

  1. C++ changes in torch/csrc/autograd/python_variable.cpp:
  - Modified the NON_TENSOR case in args processing (line 2245-2263) to check if a list contains DTensors and extract compute_mesh from them
  - Modified the NON_TENSOR case in kwargs processing (line 2305-2321) with the same logic

  2. Python changes in torch/distributed/tensor/_dispatch.py:
  - Modified list/tuple handling in args processing (line 550-576) to properly unwrap DTensors inside lists, extracting their specs for the schema and local tensors
  for execution
  - Modified list/tuple handling in kwargs processing (line 599-625) with the same unwrapping logic

  The fix ensures that:
  - The C++ dispatch layer finds the DeviceMesh from DTensors inside list arguments
  - The Python dispatch layer properly unwraps DTensors in lists before creating the OpSchema
  - Sharding propagation can proceed and correctly raise NotImplementedError for unsupported ops, regardless of whether they have list arguments

[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/_dispatch.py:
```diff
@@ -547,6 +547,33 @@ def _unwrap_to_op_info_impl(
                     )
                 )
                 local_args.append(arg)
+            elif isinstance(arg, (list, tuple)):
```
i don't love the amount of extra copypaste here.  See next PR in the stack for a refactor using a recursive helper.

zpcore replied:
I thought the pytree already flattened the args: https://github.com/pytorch/pytorch/blob/0b77806f5c9032e5d48fb1075e5eceb71317a63a/torch/distributed/tensor/_dispatch.py#L512-L517.  Is the recursive helper doing something similar?

wconstab replied:
hmm, maybe the problem is caused by user error actually: not specifying needs_pytree=True on cat registration when i should have.  although, I would prefer if it was not possible to make this mistake. I am not sure if there is a good reason to have the non-pytree path- (performance?)

wconstab committed:
Update on "[DTensor] Fix unsupported op error"


There is already an 'unsupported op' error, but it happens late enough
during sharding prop flow that for ops that take TupleStrategy
(tensorlist) inputs, we hit a different error (missing DeviceMesh from
dtensor args) which is a confusing error to tell a user when in fact the
problem is the operator is not registered with DTensor dispatch in the
first place.

This PR attempts to fix the missing devicemesh issue so that the
pre-existing error gets hit for cat as well.

Done by claude, needs careful review!!

Claude's summary:
```
 I fixed a bug where DTensor ops with list[Tensor] arguments (like torch.cat) were raising "found no DeviceMesh from dtensor args" instead of the correct "does not
  have a sharding strategy registered" error.

  Changes made:

  1. C++ changes in torch/csrc/autograd/python_variable.cpp:
  - Modified the NON_TENSOR case in args processing (line 2245-2263) to check if a list contains DTensors and extract compute_mesh from them
  - Modified the NON_TENSOR case in kwargs processing (line 2305-2321) with the same logic

  2. Python changes in torch/distributed/tensor/_dispatch.py:
  - Modified list/tuple handling in args processing (line 550-576) to properly unwrap DTensors inside lists, extracting their specs for the schema and local tensors
  for execution
  - Modified list/tuple handling in kwargs processing (line 599-625) with the same unwrapping logic

  The fix ensures that:
  - The C++ dispatch layer finds the DeviceMesh from DTensors inside list arguments
  - The Python dispatch layer properly unwraps DTensors in lists before creating the OpSchema
  - Sharding propagation can proceed and correctly raise NotImplementedError for unsupported ops, regardless of whether they have list arguments
```

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Fix unsupported op error"


There is already an 'unsupported op' error, but it happens late enough
during sharding prop flow that for ops that take TupleStrategy
(tensorlist) inputs, we hit a different error (missing DeviceMesh from
dtensor args) which is a confusing error to tell a user when in fact the
problem is the operator is not registered with DTensor dispatch in the
first place.

This PR attempts to fix the missing devicemesh issue so that the
pre-existing error gets hit for cat as well.

Done by claude, needs careful review!!

Claude's summary:
```
 I fixed a bug where DTensor ops with list[Tensor] arguments (like torch.cat) were raising "found no DeviceMesh from dtensor args" instead of the correct "does not
  have a sharding strategy registered" error.

  Changes made:

  1. C++ changes in torch/csrc/autograd/python_variable.cpp:
  - Modified the NON_TENSOR case in args processing (line 2245-2263) to check if a list contains DTensors and extract compute_mesh from them
  - Modified the NON_TENSOR case in kwargs processing (line 2305-2321) with the same logic

  2. Python changes in torch/distributed/tensor/_dispatch.py:
  - Modified list/tuple handling in args processing (line 550-576) to properly unwrap DTensors inside lists, extracting their specs for the schema and local tensors
  for execution
  - Modified list/tuple handling in kwargs processing (line 599-625) with the same unwrapping logic

  The fix ensures that:
  - The C++ dispatch layer finds the DeviceMesh from DTensors inside list arguments
  - The Python dispatch layer properly unwraps DTensors in lists before creating the OpSchema
  - Sharding propagation can proceed and correctly raise NotImplementedError for unsupported ops, regardless of whether they have list arguments
```

[ghstack-poisoned]

zpcore reviewed (approved):
LGTM!

wconstab commented:
@pytorchbot merge -i
